### PR TITLE
Pokemon Emerald: Use some new state functions, improve rule reuse

### PR DESCRIPTION
**Pull request by Zunawe**
## What is this fixing or adding?

Swaps to using `state.has_from_list_exclusive` and `state.has_group_exclusive` counters, and fixes instances where I should have just passed the bike rule in instead of making a lambda that calls it. There are other places it's correct already. Guess I just forgot and copy/pasted them around or something.

I think this technically fixes a problem where if a player starts with a badge in their starting inventory, it will get counted twice in the logic for Norman and the E4, but only once by the game itself. Nobody has ever reported this though.

Did a tiny bit of linting, but looking at what `black` thinks I should be doing, there's almost no way you're going to get me to agree to listen to pep8 100% on this file with things like

```py
    set_rule(
        get_entrance("REGION_ROUTE114/ABOVE_WATERFALL -> REGION_TERRA_CAVE_ENTRANCE/MAIN"),
        lambda state: state.has("EVENT_DEFEAT_CHAMPION", world.player)
        and state.has("TERRA_CAVE_ROUTE_114_1", world.player)
        and state.has("EVENT_DEFEAT_SHELLY", world.player),
    )
```

and

```py
            hm_rules[hm] = lambda state, hm=hm, badges=badges: state.has(
                hm, world.player
            ) and state.has_group_exclusive("Badges", world.player, badges)
```

## How was this tested?

Generated a game with local items, legendary hunt goal, and bikes enabled with a specific seed on `main` and on this branch, then compared the hash of the patched ROM. Was the same.

### CODE REVIEW

**Review by Exempt-Medic — APPROVED** [review]
Changes LGTM. Spoiler logs before and after the changes match when generating with the same seed

**Comment by Zunawe** [comment]
This and #3386 are on a collision course. Probably 3386 one will be merged first and I'll push a fix here. Just linking the PRs for completeness.

**Comment by Zunawe** [comment]
Updated to `_unique` now that 3386 merged

File: worlds/pokemon_emerald/rules.py [hunk 1]
```diff
@@ -1370,17 +1371,17 @@ def get_location(location: str):
...
     set_rule(
         get_entrance("REGION_UNDERWATER_ROUTE129/MARINE_CAVE_ENTRANCE_2 -> REGION_UNDERWATER_MARINE_CAVE/MAIN"),
-        lambda state: hm_rules["HM08 Dive"](state) and \
-            state.has("EVENT_DEFEAT_CHAMPION", world.player) and \
-            state.has("MARINE_CAVE_ROUTE_129_2", world.player) and \
-            state.has("EVENT_DEFEAT_SHELLY", world.player)
+        lambda state: hm_rules["HM08 Dive"](state)
+            and state.has("EVENT_DEFEAT_CHAMPION", world.player)
+            and state.has("MARINE_CAVE_ROUTE_129_2", world.player)
+            and state.has("EVENT_DEFEAT_SHELLY", world.player)
```
Comment (qwint) [inline worlds/pokemon_emerald/rules.py:1384]: you can do this?? i swore you had to use the `\` line breaks here

**Review by qwint — APPROVED** [review]
all changes are simple reformats besides the changes to has_group_unique("Badges") which is *technically* a change but not one that should have come up before

this was only a code review; did not run any test/gens/etc.

File: worlds/pokemon_emerald/rules.py (hunk 1 above)
Comment (Exempt-Medic) [reply worlds/pokemon_emerald/rules.py:1384]: Oh wow, I didn't think this was possible either. I usually put parens around the full thing. So like `lambda state: (thing and thing2 spanning multiple lines)

**Review by NewSoupVi — APPROVED** [review]
awesome changes, thanks for keeping your apworld up to date & cleaning it up like this :3

**Pull request merged**